<h1 style="text-align: center; font-weight: bold;">Preprocessing</h1>

In order to make the data suitable for the model, we need to preprocess it. In this notebook, we will preprocess the data in the following steps:
- Merge the datasets
- Remove duplicates
- Handle missing values
- Handle outliers
- Encode categorical features

In [39]:
import pandas as pd
import numpy as np

<h2 style="font-weight: bold;">1. Load datasets and prepare data for merging</h2>

In [40]:
# metacritic
metacritic_df = pd.read_csv("./MetaCritic/MetaCritic.csv")
print(f"MetaCritic: {metacritic_df.shape}")
print(metacritic_df.columns)

# opencritic
opencritic_df = pd.read_csv("./OpenCritic/game_data.csv")
print(f"OpenCritic: {opencritic_df.shape}")
print(opencritic_df.columns)

# steam
steam_df = pd.read_csv("./steam_scraping/data/outputs/steam_data.csv")
print(f"Steam: {steam_df.shape}")
print(steam_df.columns)

MetaCritic: (24258, 8)
Index(['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Release Date',
       'Platform', 'Rating', 'Number of Ratings'],
      dtype='object')
OpenCritic: (7771, 9)
Index(['Game Title', 'Game Genre', 'Pricing', 'Game Library Size', 'Publisher',
       'Release Date', 'Platform', 'Rating', 'Number of Rating'],
      dtype='object')
Steam: (58320, 8)
Index(['Game Title', 'Game Genre', 'Pricing', 'Publisher', 'Release Date',
       'Platform', 'Rating', 'Number of Ratings'],
      dtype='object')


There are a few columns in the datasets whose column names are not the same. We will rename these columns to make them consistent across the datasets. We will also remove the columns that are not useful for the model.

- **OpenCritic**
    - Has 'Number of Rating' instead of 'Number of Ratings' and 'Publisher' instead of 'Developer'.
    - Has an additional 'Game Library Size' column which is not needed.

- **Steam**
    - Has 'Publisher' instead of 'Developer'

- **MetaCritic** is fine as it is, so we will leave it there.

In [41]:
opencritic_df.rename(columns={"Number of Rating" : "Number of Ratings",
                              "Publisher" : "Developer"}, inplace=True)
opencritic_df.drop(columns=["Game Library Size"], inplace=True)

steam_df.rename(columns={"Publisher" : "Developer"}, inplace=True)

print(opencritic_df.columns)
print(steam_df.columns)
print(metacritic_df.columns)

Index(['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Release Date',
       'Platform', 'Rating', 'Number of Ratings'],
      dtype='object')
Index(['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Release Date',
       'Platform', 'Rating', 'Number of Ratings'],
      dtype='object')
Index(['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Release Date',
       'Platform', 'Rating', 'Number of Ratings'],
      dtype='object')


<h2 style="font-weight: bold;">2. Merging data</h2>

We will merge the datasets by `Game Title` column, which is common in all datasets.

According to dataset sizes, the priority order will be:
1. Steam (58k rows)
2. OpenCritic (24k rows)
3. MetaCritic (7k rows)

Based on the priority order, we will merge the datasets from **OpenCritic** and **MetaCritic** into **Steam**'s.
If the values in **Steam** are missing, we will fill them with the values from **OpenCritic** and **MetaCritic**.

In [42]:
# Merge dataframes
merged_df = pd.merge(steam_df, metacritic_df,
    on="Game Title",
    how="outer",
    suffixes=('_steam', '_metacritic')
)
merged_df = pd.merge(merged_df,opencritic_df,
    on="Game Title",
    how="outer",
    suffixes=('','_opencritic')
)

target_columns = ['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Release Date', 'Platform', 'Rating', 'Number of Ratings']

# Fill values based on priority
for col in target_columns[1:]:
    # strip whitespace and [, ], ' and " from the columns as well
    merged_df[col] = merged_df[f"{col}_steam"].fillna(merged_df[f"{col}_metacritic"], axis=0).fillna(merged_df[col], axis=0)
    
string_col = ['Game Title', 'Game Genre', 'Pricing', 'Developer', 'Platform', 'Release Date']
for col in string_col:
    merged_df[col] = merged_df[col].str.strip().str.strip("[").str.strip("]").str.strip("'").str.strip("'").str.replace('\u202A', '', regex=False)



merged_df = merged_df[target_columns]
merged_df

Game Title  \
0      ! That Bastard Is Trying To Steal Our Gold !   
1                                   ! Wild Russia !   
2                                        !4RC4N01D!   
3                       !4RC4N01D! 2: Retro Edition   
4                       !4RC4N01D! 2: Retro Edition   
...                                             ...   
79856                                     🔴 Circles   
79857                                     🔴 Circles   
79858                                     🔴 Circles   
79859                         🚀 Human Rocket Person   
79860                              🧠 OUT OF THE BOX   

                                              Game Genre       Pricing  \
0                       Action, Adventure, Casual, Indie         $2.99   
1                              Action, Adventure, Casual        $19.99   
2                                                 Arcade           NaN   
3                                                 Arcade           NaN   
4                                                 Arcade           NaN   
...                                                  ...           ...   
79856                                      Casual, Indie  Free To Play   
79857                                      Casual, Indie  Free To Play   
79858                                      Casual, Indie  Free To Play   
79859                          Action, Indie, Simulation         $1.99   
79860  Adventure, Casual, Indie, RPG, Simulation, Str...        $14.99   

            Developer  Release Date Platform  Rating  Number of Ratings  
0          WTFOMGames    Mar 1 2016      NaN    56.0               66.0  
1      Andreev Worlds   Apr 28 2020      NaN    61.0               60.0  
2           armogames  Jan 12, 2018       PC     4.0                4.0  
3           armogames   Feb 6, 2018       PC     3.8                4.0  
4           armogames   Feb 6, 2018       PC     3.8                4.0  
...               ...           ...      ...     ...                ...  
79856  Jeroen Wimmers   Feb 17 2017      NaN    89.0              226.0  
79857  Jeroen Wimmers   Feb 17 2017      NaN    89.0              226.0  
79858  Jeroen Wimmers   Feb 17 2017      NaN    89.0              226.0  
79859      2nd Studio   Nov 14 2018      NaN    95.0               43.0  
79860   Nuclear Tales   Jul 19 2018      NaN    88.0              195.0  

[79861 rows x 8 columns]

In [43]:
merged_df.dtypes

Game Title            object
Game Genre            object
Pricing               object
Developer             object
Release Date          object
Platform              object
Rating               float64
Number of Ratings    float64
dtype: object

<h2 style="font-weight: bold;">3. Data Cleaning</h2>

<h3 style="font-weight: bold;">3.1. Remove duplicates</h2>

As seen from the above preview, the total records of the dataset is **79861** entries. However, there are duplicates in the datasets. We will have to remove them.

In [44]:
# strip the string
merged_df['Game Title'] = merged_df['Game Title'].str.strip()
# remove duplicates
merged_df = merged_df.drop_duplicates(subset=['Game Title'])
merged_df.shape

(62373, 8)

<h3 style="font-weight: bold;">3.1. Fill or remove missing values</h2>

Since we merged 3 datasets together, there will exist rows where values from all 3 datasets are missing, hence the **NaN** values. We will fill these missing rows with accordingly.

First, we will check the missing values in the dataset.

In [45]:
def getMissingValues(df: pd.DataFrame) -> pd.Series:
    return df.isnull().sum()

missing_values = getMissingValues(merged_df)
missing_values

Game Title               0
Game Genre            2238
Pricing              13145
Developer               12
Release Date             4
Platform             41882
Rating                 600
Number of Ratings       16
dtype: int64

**Developer, Release Date and Number of Ratings** have very few missing values (<20), we can safely remove them from the dataset without significant impact.

In [ ]:
merged_df = merged_df.dropna(subset=['Developer', 'Release Date', 'Number of Ratings'])

**Game Genre** has *2238* missing values. We will fill these missing values with *mode* because it is a categorical feature.

In [47]:
# get all genre values
genres = merged_df['Game Genre'].str.split(',').explode().str.strip("[").str.strip("'").str.strip("'").str.strip("]").dropna().value_counts()

mode_genre = genres.idxmax()
merged_df.fillna({'Game Genre': mode_genre}, inplace=True)
print(f"Missing values in Game Genre replaced with mode value: {mode_genre}")

Missing values in Game Genre replaced with mode value:  Indie


**Pricing** column is skewed, so we will fill the missing values with the *median*.

But first, there are values in `Pricing` where they do not contain a numeric value (usually "Free to Play" for "Free", etc.), we will have to convert the column to a numeric type, with non-numeric (might be Free) values as 0.

In [48]:
import re

merged_df.columns
pricing_col = merged_df['Pricing']

numeric_pattern = re.compile(r'(?:\d+)+')
non_numeric_prices = pricing_col[~pricing_col.str.contains(numeric_pattern, na=False)].unique()

non_numeric_prices

array([nan, 'Free', 'Free To Play', '', 'Free to Play', 'Free Demo',
       'Free Mod', 'Play for Free!', 'Install Now', 'Play the Demo',
       'Play WARMACHINE: Tactics Demo'], dtype=object)

Now that we knew non-numeric values in the `Pricing` column are mostly *free* games, we will replace them with `0`.

We also need to convert the `Pricing` column to a numeric type. Preferably `float`.

In [49]:
# convert the non-numeric prices to 0
merged_df['Pricing'] = merged_df['Pricing'].replace(non_numeric_prices, '0')

# convert pricing to numeric
merged_df['Pricing'] = merged_df['Pricing'].str.replace(r'\$\s*', '', regex=True).astype(float)

There are still NaN values in the `Pricing` column. We will fill them with the median value since the column is skewed and does not follow a normal distribution.

In [50]:
median_pricing = merged_df['Pricing'].median()
merged_df.fillna({'Pricing' : median_pricing}, inplace=True)

**Platform** column has the most missing values, *41882*. We will fill these missing values with the *mode* since the column is categorical.

In [51]:
platforms = merged_df['Platform'].str.split(',').explode().str.strip().value_counts()
mode_platform = platforms.idxmax()
merged_df.fillna({'Platform': mode_platform}, inplace=True)
print(f"Missing values in Platform replaced with mode value: {mode_platform}")

Missing values in Platform replaced with mode value: PC


**Rating** column has *600* missing values. We will fill these missing values with *median* since the rating is skewed.

But first, we need to convert them all into numerical datatype.

In [52]:
merged_df['Rating'] = merged_df['Rating'].astype(float)
median_rating = merged_df['Rating'].median()
merged_df.fillna({'Rating': median_rating}, inplace=True)
print(f"Missing values in Rating replaced with median value: {median_rating}")

Missing values in Rating replaced with median value: 76.0


Now that we have filled/removed missing values, we will check again to make sure there are no missing values left.

In [53]:
missing_values = getMissingValues(merged_df)
missing_values

Game Title           0
Game Genre           0
Pricing              0
Developer            0
Release Date         0
Platform             0
Rating               0
Number of Ratings    0
dtype: int64

We will also check for the appropriate datatypes for each column.

In [54]:
merged_df.dtypes

Game Title            object
Game Genre            object
Pricing              float64
Developer             object
Release Date          object
Platform              object
Rating               float64
Number of Ratings    float64
dtype: object

**Number of Ratings** should be `uint64` instead of `float`.

In [55]:
merged_df['Number of Ratings'] = merged_df['Number of Ratings'].astype(np.uint64)

In [56]:
merged_df.dtypes

Game Title            object
Game Genre            object
Pricing              float64
Developer             object
Release Date          object
Platform              object
Rating               float64
Number of Ratings     uint64
dtype: object

<h2 style="font-weight: bold;">4. Save cleaned data</h2>

Sort the data alphabetically by `Game Title` and save it as a CSV file.

In [57]:
merged_df = merged_df.sort_values(by="Game Title")
merged_df.to_csv("cleaned_data.csv", index=False, encoding='utf-8')
merged_df.head()

Game Title  \
0  ! That Bastard Is Trying To Steal Our Gold !   
1                               ! Wild Russia !   
2                                    !4RC4N01D!   
3                   !4RC4N01D! 2: Retro Edition   
5                      !4RC4N01D! 3: Cold Space   

                         Game Genre  Pricing       Developer  Release Date  \
0  Action, Adventure, Casual, Indie     2.99      WTFOMGames    Mar 1 2016   
1         Action, Adventure, Casual    19.99  Andreev Worlds   Apr 28 2020   
2                            Arcade     0.00       armogames  Jan 12, 2018   
3                            Arcade     0.00       armogames   Feb 6, 2018   
5                            Arcade     0.00       armogames   Mar 8, 2018   

  Platform  Rating  Number of Ratings  
0       PC    56.0                 66  
1       PC    61.0                 60  
2       PC     4.0                  4  
3       PC     3.8                  4  
5       PC     3.0                  4